### PR TITLE
QueryFunctionsDescriptor returns all fields for multi-fielded functions

### PR DESCRIPTION
**Pull request by apmoriarty**
Addresses https://github.com/NationalSecurityAgency/datawave/issues/2445

- added unit test for QueryFunctionsDescriptor
- corrected bug that prevented all fields from being returned in a multi-fielded function
- added test cases to QueryOptionsFromQueryVisitorTest

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java]: Read the function description.  The arg argument is to be used to denote argument we are to be returning the fields for normalization.  So for example if the first argument denotes the name of the fields, then for arg=0, this should return the empty set because we are not normalizing the field names themselves.

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java]: Got it, I'll revert that change. Do any additional functions need to delegate to the fields method besides `includeText` (such as `matchRegex`) ?

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java
Comment (mineralntl) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java:158]: Just for clarification - does it matter that we're doing a simple addAll(identifier) here but down below we're doing a UniqueFields.from(identifier)?

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java:158]: No, it does not matter. These functions will all have arguments that are ASTIdentifiers. The #UNIQUE function might have arguments that are either an ASTIdentifier OR an ASTStringLiteral.

Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java:158]: 👍

**Review by mineralntl — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java]: So the idea was that the "fields" being returned are the the fields for which thier normalizers should be applied to the argument in question.  Originally all of the methods being handled by the descriptor essentially would need to normalize all arguments past the first one using all of the fields, except for include_text which was explicitly made to bypass any normalization.
So to answer your question, look at the arguments for all methods being handled by the QueryFunctions and see if that holds.  If not then we may need to adjust something.

Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java]: After some poking around I don't believe matchRegex needs to be added to the exclusion list.

**Review by alerman — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java
Comment (jwomeara) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/QueryFunctionsDescriptor.java]: Inconsistent with the other cases.  either do a static import for all terms, or none.

**Review by jwomeara — COMMENTED** [review]
looks safe from a query microservices perspective.  happy to approve from a codeowner perspective once this has 2 approvals.

**Review by mineralntl — APPROVED** [review]

**Review by hlgp — APPROVED** [review]

**Review by jwomeara — APPROVED** [review]

**Pull request merged**
